src/generator.rs

```diff
@@ -0,0 +1,239 @@
+            // do nothing
+        }
+    }
+
+    let waker = Waker::from(Arc::new(NoopWake));
+    let mut context = Context::from_waker(&waker);
+
+    match task.as_mut().poll(&mut context) {
+        Poll::Pending => None,
+        Poll::Ready(item) => Some(item),
+    }
+}
+
+/// Utility types and methods
+impl<'a, YieldTy, ResumeTy, OutTy> Generator<'a, YieldTy, ResumeTy, Result<OutTy, crate::Error>>
+where
+    OutTy: Send + 'a,
+{
+    pub fn resolve_to_result(&mut self) -> Result<OutTy, crate::Error> {
+        let state = self.start();
+        match state {
+            GeneratorState::Suspended(_) => Err(Error::new(
+                crate::ErrorKind::UnsupportedFunction,
+                "cannot finish generator",
+            )),
+            GeneratorState::Completed(res) => res,
+        }
+    }
+
+    pub fn unwrap(&mut self) -> OutTy {
+        self.resolve_to_result().unwrap()
+    }
+
+    pub fn expect(&mut self, msg: &str) -> OutTy {
+        self.resolve_to_result().expect(msg)
+    }
+}
+#[derive(Debug)]
+pub struct NetworkRequest {
+    pub protocol: NetworkProtocol,
+    pub url: Url,
+    pub data: Vec<u8>, // avoid life time problem, suspend requires 'static life time
+}
+
+impl<'a, YieldTy, ResumeTy, OutTy> std::fmt::Debug for Generator<'a, YieldTy, ResumeTy, OutTy>
+where
+    YieldTy: std::fmt::Debug,
+    ResumeTy: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Generator")
+            .field("yielded_value", &self.yielded_value)
+            .field("resumed_value", &self.resumed_value)
+            .finish()
+    }
+}
+pub type GeneratorInitSecurityContext<'a> =
+    Generator<'a, NetworkRequest, crate::Result<Vec<u8>>, crate::Result<InitializeSecurityContextResult>>;
+
+pub type GeneratorChangePassword<'a> = Generator<'a, NetworkRequest, crate::Result<Vec<u8>>, crate::Result<()>>;
+
+pub(crate) type YieldPointLocal = YieldPoint<NetworkRequest, crate::Result<Vec<u8>>>;
+
+impl<'a, YieldType, ResumeType, OutType, ErrorType> From<Result<OutType, ErrorType>>
+    for Generator<'a, YieldType, ResumeType, Result<OutType, ErrorType>>
+where
+    OutType: Send + 'a,
+    ErrorType: Send + 'a,
+{
+    fn from(value: Result<OutType, ErrorType>) -> Self {
+        Generator::new(move |_| async move { value })
+    }
+}
+/// Utilitiies for working with network client
+#[cfg(feature = "network_client")]
+pub trait ResolveWithDefaultNetWorkClient<'a, OutTy>
+where
+    OutTy: 'a + Send,
+{
+    fn resolve_with_default_network_client(&mut self) -> OutTy;
+}
+
+#[cfg(feature = "network_client")]
+impl<'a, OutTy> ResolveWithDefaultNetWorkClient<'a, OutTy>
+    for Generator<'a, NetworkRequest, crate::Result<Vec<u8>>, OutTy>
+where
+    OutTy: 'a + Send,
+{
+    fn resolve_with_default_network_client(&mut self) -> OutTy {
+        let network_client = crate::network_client::reqwest_network_client::ReqwestNetworkClient {};
```

Comment:
Nit: `{}` are not required since this is a "unit" struct
```suggestion
        let network_client = crate::network_client::reqwest_network_client::ReqwestNetworkClient;
```